### DISCUSSION TITLE
Vercel production deployment is not running custom cron job properly.

### DISCUSSION THREAD

**Discussion by Gukkey** [vercel/community#7370] [Help]
### Summary

Hi, I don't know whether this is an next js issue or vercel issue but here it is. So I have a project that runs a [cron job (which is built using npm cron package)](https://github.com/Gukkey/project-forum/blob/main/src/app/api/cron/cron.ts) whenever a post or get [request is sent](https://github.com/Gukkey/project-forum/blob/main/src/app/api/cron/route.ts) by QStash schedules. I can tell you it works well when I run it as a development build ( using `npm run dev`), even during a production build (using `npm run build && npm start`).

![Code_twjaoACwEY](https://github.com/vercel/vercel/assets/74915094/2d8fb5d6-fd73-43e4-a30a-238174766ae1)
![Code_2hFImm91up](https://github.com/vercel/vercel/assets/74915094/8bf2e716-323e-4d2a-a043-95f82e5f84a2)
![Bruno_4QnM3kk86d](https://github.com/vercel/vercel/assets/74915094/5c942f1e-b95d-44cb-95cc-1562cc03b290)
![Arc_wKKkNbpqVl](https://github.com/vercel/vercel/assets/74915094/dcb6cbe9-599d-4f44-93ee-a459a0e92ab8)

But where it does not run is vercel deployment targeted at our main branch. I mean it runs but no log is being stored in the database, or the other infomation that would be stored in the database if the cron job runs correctly. I have double checked environment variables, tested it multiple times in development build (there are some bug fixes that needed to be done in the backend but it still works perfectly in the front end in both development and production build).

![Bruno_yHzQU1brB5](https://github.com/vercel/community/assets/74915094/13a5c445-62b1-4c09-9108-0a18d309353a)
![Arc_McJCRn7lih](https://github.com/vercel/community/assets/74915094/7f8e6774-4ea1-497f-ac22-f14bbe4f3fec)
p.s) the second record saying "cron job completed" is from the cron job that we invoked by development build

In these images you can see that the our project receives the post request from the Qstash and returns a response properly
![Arc_WSKOZeWUbr](https://github.com/vercel/community/assets/74915094/ee21353a-ae42-4f14-aef1-6612c2098169)

So why does cron job is not working properly or atleast the data is not stored in the database properly? Even if there is an error, vercel would always restart the build in such cases right? I also made sure to log things before that happens too. 

### Example

Production deployment: https://moody-blues.vercel.app/api/cron (other routes are authenticated)

### Steps to Reproduce

It will be painful to replicate but here it is. 

1. Clone the repo, add the necessary variables from `.env.local.example`.
2. Go to Upstash's Qstash and create a schedule for your deployment. If your production deployment is `https://abcd.app`, make a schedule to `https://abcd.app/api/cron` with custom timings `*/15 * * * *`. This will send a post request to invoke your cron job every 15 minutes by itself. 
3. Go to Upstash's Redis and create a redis database. Their token and url would go to `REDIS_URL` and 
`REDIS_TOKEN`. Also create a database in vercel, and then push the tables in the project by `npx drizzle-kit push`
4. Go to anichart.net and head to the nearest releasing anime and grab some of their ids and add them to `CRON_JOB_ANIME_ID_ARRAY`
5. As for `CRON_JOB_SECTION_ID`, `CRON_JOB_TOPIC_ID`, `CRON_JOB_USER_ID`, you have to run drizzle studio (by `npx drizzle-kit studio`) and then head to section and topic tables and create them by adding new records. For user ids, signin/signup is recommended. You have to add clerk credentials for them. 
![image](https://github.com/vercel/community/assets/74915094/99944e18-91c9-4235-932d-3bbe2ffbc4ec)
![image](https://github.com/vercel/community/assets/74915094/6631d8c7-2bfc-4fec-b4ed-e90138202d8c)
![image](https://github.com/vercel/community/assets/74915094/51667969-188d-4ef3-9820-0bd135eccdac)
6. Other env variables mentioned in the projects are required as well.

**Comment by pawlean** [comment]
Hi @Gukkey!

Some things to check and try out:

1. Ensure all required environment variables are correctly set in production.
2. Check for logs in production to verify if the cron job is triggered and executed properly.
3. Confirm that all tokens, keys, and URLs used by the cron job are valid in production.
4. Compare development and production environments for differences in configurations or dependencies.
5. Ensure the cron job is included and correctly configured in the Vercel deployment settings.
6. Verify that the production environment can connect to and write to the database.
7. Implement error handling and retry mechanisms in the cron job code.
8. Set up monitoring and alerts to detect and notify about cron job failures.
9. Confirm that QStash schedules are correctly set to trigger the cron job in production.
10. Ensure there are no code differences between the development and production branches affecting the cron job.

[a comment by an automated account is not shown]
